cortex-ar/src/mmu.rs

```diff
@@ -0,0 +1,167 @@
+#[derive(Debug, Copy, Clone)]
+pub struct MemoryRegionAttributesRaw {
+    /// TEX bits
+    type_extensions: u8,
+    c: bool,
+    b: bool,
+}
+
+impl MemoryRegionAttributesRaw {
+    pub const fn new(type_extensions: u8, c: bool, b: bool) -> Self {
+        Self {
+            type_extensions,
+            c,
+            b,
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum CacheableMemoryAttribute {
+    NonCacheable = 0b00,
+    WriteBackWriteAlloc = 0b01,
+    WriteThroughNoWriteAlloc = 0b10,
+    WriteBackNoWriteAlloc = 0b11,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum MemoryRegionAttributes {
+    StronglyOrdered,
+    ShareableDevice,
+    OuterAndInnerWriteThroughNoWriteAlloc,
+    OuterAndInnerWriteBackNoWriteAlloc,
+    OuterAndInnerNonCacheable,
+    OuterAndInnerWriteBackWriteAlloc,
+    NonShareableDevice,
+    CacheableMemory {
+        inner: CacheableMemoryAttribute,
+        outer: CacheableMemoryAttribute,
+    },
+}
+
+impl MemoryRegionAttributes {
+    pub const fn as_raw(&self) -> MemoryRegionAttributesRaw {
+        match self {
+            MemoryRegionAttributes::StronglyOrdered => {
+                MemoryRegionAttributesRaw::new(0b000, false, false)
+            }
+            MemoryRegionAttributes::ShareableDevice => {
+                MemoryRegionAttributesRaw::new(0b000, false, true)
+            }
+            MemoryRegionAttributes::OuterAndInnerWriteThroughNoWriteAlloc => {
+                MemoryRegionAttributesRaw::new(0b000, true, false)
+            }
+            MemoryRegionAttributes::OuterAndInnerWriteBackNoWriteAlloc => {
+                MemoryRegionAttributesRaw::new(0b000, true, true)
+            }
+            MemoryRegionAttributes::OuterAndInnerNonCacheable => {
+                MemoryRegionAttributesRaw::new(0b001, false, false)
+            }
+            MemoryRegionAttributes::OuterAndInnerWriteBackWriteAlloc => {
+                MemoryRegionAttributesRaw::new(0b001, true, true)
+            }
+            MemoryRegionAttributes::NonShareableDevice => {
+                MemoryRegionAttributesRaw::new(0b010, false, false)
+            }
+            MemoryRegionAttributes::CacheableMemory { inner, outer } => {
+                MemoryRegionAttributesRaw::new(
+                    (1 << 2) | (*outer as u8),
+                    (*inner as u8 & 0b10) != 0,
+                    (*inner as u8 & 0b01) != 0,
+                )
+            }
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct SectionAttributes {
+    /// NG bit
+    pub non_global: bool,
+    /// Implementation defined bit.
+    pub p_bit: bool,
+    pub shareable: bool,
+    /// AP bits
+    pub access: AccessPermissions,
+    pub memory_attrs: MemoryRegionAttributesRaw,
+    pub domain: u8,
+    /// xN bit.
+    pub execute_never: bool,
+}
+
+/// 1 MB section translation entry, mapping a 1 MB region to a physical address.
+#[derive(Debug, Copy, Clone)]
+pub struct L1Section(pub u32);
```

Comment:
perhaps a bitbybit::bitfield is useful here?

Reply:
could be useful. I never used the crate before, I can have a look at it.